File: 搜索算法/二分查找/BinarySearch.java
import java.util.*;

public class BinarySearch {
    public static <T extends Comparable> int[]
            binarySearch(List<T> array, T target) {
        if (array == null || target == null)
            throw new RuntimeException(
                "array == null || target == null");

        int low = 0, high = array.size() - 1;
        int[] result = new int[2];
        while (low <= high) {
            int middle = (low + high) / 2;
            if (array.get(middle).compareTo(target) == 0) {
                result[0] = result[1] = middle;
                return result;
            }

            if (array.get(middle).compareTo(target) == 1) {
                if (middle == low ||
                        array.get(middle-1).compareTo(target) == -1) {
                    result[0] = middle - 1;
                    result[1] = middle;
                    return result;
                }
                high = middle - 1; 
                continue;
            }

            if (middle == array.size() - 1 ||
                    array.get(middle+1).compareTo(target) == 1) {
                result[0] = middle; result[1] = middle + 1;
                return result;
            }
            low = middle + 1;
        }

        // unreachable
        return result;
    }

    public static void main(String[] args) {
        List<Integer> array = Arrays.asList(new Integer[]{1, 2, 4, 7, 8, 9, 10});
        int[] result;
        result = binarySearch(array, 11);
        System.out.println(String.format(
            "(%d, %d)", result[0], result[1]));
        result = binarySearch(array, 9);
        System.out.println(String.format(
            "(%d, %d)", result[0], result[1]));
        result = binarySearch(array, 5);
        System.out.println(String.format(
            "(%d, %d)", result[0], result[1]));
        result = binarySearch(array, 1);
        System.out.println(String.format(
            "(%d, %d)", result[0], result[1]));
        result = binarySearch(array, 0);
        System.out.println(String.format(
            "(%d, %d)", result[0], result[1]));
    }
}

File: 搜索算法/二分查找/BinarySearch.py
def binary_search_recursive(array, target, low=None, high=None):
    if low is None:
        low = 0
    if high is None:
        high = len(array) - 1

    if low > high:
        return -1

    mid = (low + high) / 2
    if target == array[mid]:
        return mid
    if target < array[mid]:
        return binary_search_recursive(array, target, low, mid - 1)
    return binary_search_recursive(array, target, mid + 1, high)


def binary_search(array, target):
    low = 0
    high = len(array) - 1

    while low <= high:
        mid = (low + high) / 2
        if target == array[mid]:
            return mid
        elif target < array[mid]:
            high = mid - 1
        else:
            low = mid + 1

    return -1


def test():
    array = list(range(20))
    for ind, element in enumerate(array):
        assert binary_search(array, element) == ind
        assert binary_search_recursive(array, element) == ind
    assert binary_search(array, -1) == -1


if __name__ == "__main__":
    test()
